### ISSUE TITLE
APQ GET requests often (but inconsistently) resulting in "provided sha does not match query"

### ISSUE THREAD

**Issue by sammous1** [FormidableLabs/urql#2336]
Hi,

I'm trying to leverage the `@urql/exchange-persisted-fetch` to implement APQs within my application. I set up an urql client that includes the `persistedFetchExchange` in its list of exchanges. When I make queries using `useQuery`, I conditionally override some context that gets sent with the request, such that the `context.url` gets altered to a different endpoint and `context.preferGetMethod` gets set to `true`. The latter is normally set to `false` as our queries are large enough that they need to be sent as POSTs, but several can be switched over to GETs on a query-by-query basis.

For all POST requests, they succeed consistently, successfully carrying out the APQ flow. However, when I try to switch some queries over to GETs with the different URL, they fail with a 415 error. After adding the `context.fetchOptions.headers: { ['Content-Type']: 'application/json' }`, they then begin failing with "provided sha does not match query".

Some investigation I've tried:
- Provided `generateHash` arg to APQ exchange that uses same [sha256 implementation as apollo server](https://github.com/apollographql/apollo-server/blob/6954aff9b56bbb8eb0b60928df2c3e8a7941b459/packages/apollo-server-core/src/utils/createSHA.ts)
- Checked the hashes being sent for the same query with modified methods (GETs vs POSTs)/headers (including or excluding the content type) and they all look to be the same. The times the query did succeed, they also had the same hash.
- It looked like an additional newline character was being sent in the query so I also trimming that out in the `generateHash` method, but that also failed.

urql version & exchanges:
```
"urql": "^2.2.0",
"@urql/devtools": "^2.0.0",
"@urql/exchange-persisted-fetch": "^1.3.4",
```

**Steps to reproduce**
See above.

**Expected behavior**
Both POST and GET requests should consistently succeed following the APQ flow.

**Actual behavior**
POST requests always succeed, while GET requests usually fail (but will sometimes succeed).

**Comment by kitten** [comment]
Hm, first of all, thanks for the investigation notes! This means we can already rule out an inconsistent SHA encoding here. However, if they _look_ the same, we still have to look into why the generation here differs. I'm assuming that the query we're sending isn't arriving at the same SHA _server-side_. So that means we may be missing a crucial step here, or there's a slight difference in how the server-side checks the query and how we arrive at the SHA.

I'm actually thinking it may be this formatting here, potentially, which we haven't updated with the new formatting that `request.ts` already uses. https://github.com/FormidableLabs/urql/blob/8e92357c5d1a7db2e28827377bda4e781dcd8e3b/packages/core/src/internal/fetchOptions.ts#L43

Do you have a sample query that fails with a GET request? I'm suspecting it'll either contain a comment that isn't formatted correctly or it maybe fails to use a variable for a string that contains a `#` character. Alternatively, it's obviously also possible that Apollo Server doesn't re-print the query before hashing, which would be ... problematic for GET requests 😆 

However, since this only happens for `GET` methods, I'm thinking it's likely the line I've referenced. If it's happening _inconsistently_ then likely removing newlines and such is OK, but it may be related to this being a heuristic that we'll have to match with: https://github.com/FormidableLabs/urql/blob/8e92357c5d1a7db2e28827377bda4e781dcd8e3b/packages/core/src/utils/request.ts#L30-L39

**Comment by sammous1** [comment]
Thanks for the quick response! Here's a sample failing request URL. Note: the hash won't be correct since I tried to "sanitize" the URL/query to redact some data at the advice of coworkers.



https://someGraphQLServer.com/?operationName=opName&query=query%20opName(%24l%3A%20Int!%2C%20%24pToken%3A%20String%2C%20%24qF%3A%20sQF!)%20%7B%0A%20%20pT(l%3A%20%24l%2C%20pToken%3A%20%24pToken%2C%20qF%3A%20%24qF)%20%7B%0A%20%20%20%20ti%20%7B%0A%20%20%20%20%20%20lT%20%7B%0A%20%20%20%20%20%20%20%20...TVM%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20...TTM%0A%20%20%20%20%7D%0A%20%20%20%20pToken%0A%20%20%7D%0A%7D%0A%0Afragment%20TTM%20on%20T%20%7B%0A%20%20id%0A%20%20tT%20%7B%0A%20%20%20%20tx%0A%20%20%7D%0A%20%20pI%20%7B%0A%20%20%20%20id%0A%20%20%20%20w%0A%20%20%20%20h%0A%20%20%20%20u%0A%20%20%7D%0A%20%20rD%20%7B%0A%20%20%20%20d%0A%20%20%20%20m%0A%20%20%20%20y%0A%20%20%7D%0A%7D%0A%0Afragment%20TVM%20on%20V%20%7B%0A%20%20id%0A%20%20n%20%7B%0A%20%20%20%20v%0A%20%20%7D%0A%20%20rT%20%7B%0A%20%20%20%20v%0A%20%20%7D%0A%20%20des%20%7B%0A%20%20%20%20v%0A%20%20%7D%0A%20%20thu%20%7B%0A%20%20%20%20u%0A%20%20%20%20w%0A%20%20%20%20h%0A%20%20%7D%0A%7D&variables=%7B%22l%22%3A100%2C%22qF%22%3A%7B%22rDR%22%3A%7B%22end%22%3A%222022-06-07%22%2C%22start%22%3A%222022-03-09%22%7D%7D%7D&extensions=%7B%22persistedQuery%22%3A%7B%22sha256Hash%22%3A%229f030f76b816054ffcc1a37f6ee49efbc7a4e15b3cd24a29fffdc60e5321037b%22%2C%22version%22%3A1%7D%7D

Decoded in the browser:
```
operationName: opName
query: query opName($l: Int!, $pToken: String, $qF: sQF!) {
  pT(l: $l, pToken: $pToken, qF: $qF) {
    ti {
      lT {
        ...TVM
      }
      ...TTM
    }
    pToken
  }
}

fragment TTM on T {
  id
  tT {
    tx
  }
  pI {
    id
    w
    h
    u
  }
  rD {
    d
    m
    y
  }
}

fragment TVM on V {
  id
  n {
    v
  }
  rT {
    v
  }
  des {
    v
  }
  thu {
    u
    w
    h
  }
}
variables: {"l":100,"qF":{"rDR":{"end":"2022-06-07","start":"2022-03-09"}}}
extensions: {"persistedQuery":{"sha256Hash":"9f030f76b816054ffcc1a37f6ee49efbc7a4e15b3cd24a29fffdc60e5321037b","version":1}}
```

**Comment by sammous1** [comment]
Prior to redacting some stuff, I don't think it contains a comment that isn't formatted correctly or fails to use a variable for a string that contains a # character.

**Comment by sammous1** [comment]
Actually, it looks like that trailing newline character in the query might be the culprit. I updated my `generateHash` argument in my `persistedFetchExchange` to be the following method:
```
function computeQueryHash(query: string) {
    let queryToHash = query;
    if ([
        'failingGETQueryName1`',
        'failingGETQueryName2',
        'failingGETQueryName3'
    ].some(queryName => query.includes(queryName))) {
        queryToHash = query.trim();
    }
    return require('sha.js')('sha256').update(queryToHash).digest('hex');
}
```

And they started working consistently. I had forgotten that when I tried to trim the newline before, it caused the POST requests to start failing, so I abandoned it. But revisiting it, I can see that it helps get the GET queries to hash correctly and the pages to load properly if I conditionally do this trim.

**Comment by kitten** [comment]
I _may_ have an idea what the culprit here is. Have you checked whether your server-side is on the same graphql major version compared to the client-side? I believe printing has slightly changed recently in `graphql@16` and the trailing newline is now being omitted by it, which could cause a client to server mismatch.
Unfortunately that wouldn't explain why it'd only happen sometimes

**Comment by sammous1** [comment]
Good idea, but unfortunately I don't think the MV mismatch is the cause. It was `"graphql": "^14.6.0"` on the client and `"graphql": "^15.6.0"` on the server, but after bumping the client up to the same, I still saw the "provided sha does not match query" error.

**Comment by sammous1** [comment]
Since you mentioned that change about `graphql@16`, I tried bumping up my client to "graphql": "^16.3.0" (while leaving the server at `"graphql": "^15.6.0"`) and things seem to be working; thanks for that tip!

**Comment by sammous1** [comment]
Actually still facing issues here. My application also leverages urql's ssrExchange, and I'm periodically getting the error `SyntaxError: Unexpected token o in JSON at position 1`, which looks like it's thrown from here: https://github.com/FormidableLabs/urql/blob/c4a0edebe9bf318ab1e97f5e5ec0349513eb4e73/packages/core/src/exchanges/ssr.ts#L75. 

Was using the above graphql versions, but concerned that mismatch might be causing this issue now.

**Comment by JoviDeCroock** [comment]
@sammous1 that would imply that your `result.data` is just a string starting with o, any chance you could log out what the data is there. Maybe your API isn't returning correctly at times. Does look unrelated to the original issue though

**Comment by sammous1** [comment]
I think that error message was misleading; I adjusted the urql code to include some console logs, and found that `result.data` contained an object, not a string. Any ideas why that may have been the case? Seems like it should never be the case given the typing.

**Comment by sammous1** [comment]
Update: I think I've found the root cause there too. We have a custom cache exchange implementation for the `urqlCache` property whose `serialize` implementation does not stringify the result from the server. Updating that seems to have solved the issue.
